### DISCUSSION TITLE
mobile-sdk-android build error

### DISCUSSION THREAD

**Discussion by paprophet** [iTwin/itwinjs-core#4849] [Q&A]
https://github.com/iTwin/mobile-sdk-android

git clone https://github.com/iTwin/mobile-native-android
cd mobile-native-android
./gradlew --no-daemon assembleGitHub publishToMavenLocal

BUILD FAILED: 
Administrator@Win10-2022QPPJR MINGW64 /c/gitcode/mobile-native-android (main)
$ ./gradlew.bat --no-daemon assembleGitHub publishToMavenLocal --info
Initialized native services in: C:\Users\Administrator\.gradle\native
Initialized jansi services in: C:\Users\Administrator\.gradle\native
To honour the JVM settings for this build a single-use Daemon process will be forked. See https://docs.gradle.org/7.4/userguide/gradle_daemon.html#sec:disabling_the_daemon.
Starting process 'Gradle build daemon'. Working directory: C:\Users\Administrator\.gradle\daemon\7.4 Command: C:\Program Files\Java\jdk1.8.0_202\bin\java.exe -XX:MaxMetaspaceSize=256m -XX:+HeapDumpOnOutOfMemoryError -Xms256m -Xmx512m -Dfile.encoding=GBK -Duser.country=CN -Duser.language=zh -Duser.variant -cp C:\Users\Administrator\.gradle\wrapper\dists\gradle-7.4-bin\c0gwcg53nkjbqw7r0h0umtfvt\gradle-7.4\lib\gradle-launcher-7.4.jar org.gradle.launcher.daemon.bootstrap.GradleDaemon 7.4
Successfully started process 'Gradle build daemon'
An attempt to start the daemon took 5.408 secs.
The client will now receive all logging from the daemon (pid: 3348). The daemon log file: C:\Users\Administrator\.gradle\daemon\7.4\daemon-3348.out.log
Daemon will be stopped at the end of the build
Using 12 worker leases.
Watching the file system is configured to be enabled if available
Now considering [C:\gitcode\mobile-native-android] as hierarchies to watch
File system watching is active
Starting Build
Settings evaluated using settings file 'C:\gitcode\mobile-native-android\settings.gradle'.
Projects loaded. Root project using build file 'C:\gitcode\mobile-native-android\build.gradle'.
Included projects: [root project 'mobile-native-android']

> Configure project :
Evaluating root project 'mobile-native-android' using build file 'C:\gitcode\mobile-native-android\build.gradle'.
All projects evaluated.
Selected primary task 'assembleGitHub' from project :
Selected primary task 'publishToMavenLocal' from project :
Tasks to be executed: [task ':assembleGitHub', task ':generatePomFileForMavenPublication', task ':publishMavenPublicationToMavenLocal', task ':publishToMavenLocal']
Tasks that were excluded: []
:assembleGitHub (Thread[Execution worker for ':' Thread 5,5,main]) started.

> Task :assembleGitHub FAILED
Caching disabled for task ':assembleGitHub' because:
  Build cache is disabled
Task ':assembleGitHub' is not up-to-date because:
  Task has not declared any outputs despite executing actions.
Starting process 'command 'gh''. Working directory: C:\gitcode\mobile-native-android Command: gh release download 3.6.0 -p iTwinAndroidLibrary.aar
:assembleGitHub (Thread[Execution worker for ':' Thread 5,5,main]) completed. Took 0.195 secs.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':assembleGitHub'.
> A problem occurred starting process 'command 'gh''

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 18s
1 actionable task: 1 executed
Watched directory hierarchies: []

**Comment by pmconne** [comment]
Try installing [GitHub client](https://github.com/cli/cli#installation) and restarting your shell.
Heed the warning:
> This is pre-release software and provided as-is.

@tcobbs-bentley or @toddsouthenbentley may be able to provide more detailed guidance but responses may be delayed due to holidays.

**Reply by paprophet** [reply]
ok ,i installing [GitHub client](https://github.com/cli/cli#installation) and restarting your shell.
but in mobile-sdk-android and run gradlew.bat 

Administrator@Win10-2022QPPJR MINGW64 /c/gitcode/mobile-sdk-android (main)
$ ./gradlew.bat --no-daemon publishToMavenLocal --info

* What went wrong:
Execution failed for task ':mobile-sdk:generateDebugRFile'.
> Could not resolve all files for configuration ':mobile-sdk:debugCompileClasspath'.
   > Could not resolve com.github.itwin:mobile-native-android:3.4.21.
     Required by:
         project :mobile-sdk
      > Could not resolve com.github.itwin:mobile-native-android:3.4.21.
         > Could not get resource 'https://maven.pkg.github.com/itwin/mobile-native-android/com/github/itwin/mobile-native-android/3.4.21/mobile-native-android-3.4.21.pom'.
            > Could not GET 'https://maven.pkg.github.com/itwin/mobile-native-android/com/github/itwin/mobile-native-android/3.4.21/mobile-native-android-3.4.21.pom'. Received status code 401 from server: Unauthorized

**Reply by tcobbs-bentley** [reply]
You must create a GitHub Personal Access Token (PAT). That's documented in our [setup instructions for our Android samples](https://github.com/iTwin/mobile-samples/blob/main/Android/README.md), but not here. I will fix that by adding it to CONTRIBUTING.md.

**Reply by tcobbs-bentley** [reply]
I have now updated the README.md for iTwin/mobile-sdk-android to link to its CONTRIBUTING.md, and I have updated the CONTRIBUTING.md to note that the GitHub CLI is required, as well as noting that a GitHub PAT is required. Hopefully this clears up confusion about how to locally build iTwin/mobile-sdk-android.

**Comment by tcobbs-bentley** [comment]
@paprophet: @pmconne's solution should fix the problem. We will update [CONTRIBUTING.md](https://github.com/iTwin/mobile-sdk-android/blob/main/CONTRIBUTING.md) to explicitly state that you must install the gh CLI. (It already mentions that that is used.) We'll also update [README.md](https://github.com/iTwin/mobile-sdk-android/blob/main/README.md) to mention CONTRIBUTING.md.

**Comment by arnab-git-hub** [comment]
@paprophet Has the solution provided resolved the issue? If yes, can you please mark it as answered. Thank you.
